`Julie Arand`

# PyCity Schools Analysis

* As a whole, schools with higher budgets, did not yield better test results. By contrast, schools with higher spending 645-675 per student actually underperformed compared to schools with smaller budgets (585 per student).

* As a whole, smaller and medium sized schools dramatically out-performed large sized schools on passing math performances (89-91% passing vs 67%).

* As a whole, charter schools out-performed the public district schools across all metrics. However, more analysis will be required to glean if the effect is due to school practices or the fact that charter schools tend to serve smaller student populations per school. 
---

In [1]:
# Dependencies and Setup
import pandas as pd
from pathlib import Path

# File to Load (Remember to Change These)
path_of_school_data  = Path("Resources/schools_complete.csv")
path_of_student_data = Path("Resources/students_complete.csv")

# Read School and Student Data File and store into Pandas DataFrames
school_data = pd.read_csv(path_of_school_data)
student_data = pd.read_csv(path_of_student_data)

# Combine the data into a single dataset.  
school_data_complete = pd.merge(student_data, school_data, how="left", on=["school_name", "school_name"])
school_data_complete.head()

Student ID       student_name gender grade        school_name  \
0           0       Paul Bradley      M   9th  Huang High School   
1           1       Victor Smith      M  12th  Huang High School   
2           2    Kevin Rodriguez      M  12th  Huang High School   
3           3  Dr. Richard Scott      M  12th  Huang High School   
4           4         Bonnie Ray      F   9th  Huang High School   

   reading_score  math_score  School ID      type  size   budget  
0             66          79          0  District  2917  1910635  
1             94          61          0  District  2917  1910635  
2             90          60          0  District  2917  1910635  
3             67          58          0  District  2917  1910635  
4             97          84          0  District  2917  1910635

## District Summary

In [2]:
# Calculate the total number of unique schools
school_count = len(school_data_complete.school_name.unique())
school_count

15

In [3]:
# Calculate the total number of students
student_count = len(school_data_complete.student_name)
student_count

39170

In [4]:
# Calculate the total budget
total_budget = school_data_complete.budget.sum()
total_budget

82932329558

In [5]:
# Calculate the average (mean) math score
average_math_score = round(school_data_complete.math_score.mean(), 2)
average_math_score

78.99

In [6]:
# Calculate the average (mean) reading score
average_reading_score = round(school_data_complete.reading_score.mean(), 2)
average_reading_score

81.88

In [7]:
# Use the following to calculate the percentage of students who passed math (math scores greather than or equal to 70)
passing_math_count = school_data_complete[school_data_complete["math_score"] >= 70].count()["student_name"]
passing_math_percentage = round(passing_math_count / float(student_count) * 100, 2)
passing_math_percentage

74.98

In [8]:
# Calculate the percentage of students who passeed reading (hint: look at how the math percentage was calculated)  
passing_reading_count = school_data_complete[school_data_complete["reading_score"] >= 70].count()["student_name"]
passing_reading_percentage = round(passing_reading_count / float(student_count) * 100, 2)
passing_reading_percentage

85.81

In [9]:
# Use the following to calculate the percentage of students that passed math and reading
passing_math_reading_count = school_data_complete[(school_data_complete["math_score"] >= 70)
                                                  & (school_data_complete["reading_score"] >= 70)].count()["student_name"]
overall_passing_rate = round(passing_math_reading_count /  float(student_count) * 100, 2)
overall_passing_rate

65.17

In [10]:
# Create a high-level snapshot of the district's key metrics in a DataFrame
district_summary = pd.DataFrame({"Total Schools": school_count,
                                 "Total Students": f"{student_count:,}",
                                 "Total Budget": f"${total_budget:,.2f}",
                                 "Average Math Score": f"{average_math_score:.6f}",
                                 "Average Reading Score": f"{average_reading_score:.5f}",
                                 "% Passing Math": f"{passing_math_percentage:.6f}",
                                 "% Passing Reading": f"{passing_reading_percentage:.6f}",
                                 "% Overall Passing": f"{overall_passing_rate: .6f}"}, index=[0])

# Display the DataFrame
district_summary

Total Schools Total Students        Total Budget Average Math Score  \
0             15         39,170  $82,932,329,558.00          78.990000   

  Average Reading Score % Passing Math % Passing Reading % Overall Passing  
0              81.88000      74.980000         85.810000         65.170000

## School Summary

- Create an overview table that summarizes key metrics about each school, including:

    - School Name
    - School Type
    - Total Students
    - Total School Budget
    - Per Student Budget
    - Average Math Score
    - Average Reading Score
    - % Passing Math
    - % Passing Reading
    - % Overall Passing (The percentage of students that passed math and reading.)

- Create a dataframe to hold the above results

In [11]:
# Use the code provided to select the school type

school_name = school_data_complete.set_index('school_name').groupby(['school_name'])

In [12]:
# school types by school name

school_type = school_data.set_index('school_name')['type']

In [13]:
#  Calculate total students

total_student = school_name['Student ID'].count()

total_student

school_name
Bailey High School       4976
Cabrera High School      1858
Figueroa High School     2949
Ford High School         2739
Griffin High School      1468
Hernandez High School    4635
Holden High School        427
Huang High School        2917
Johnson High School      4761
Pena High School          962
Rodriguez High School    3999
Shelton High School      1761
Thomas High School       1635
Wilson High School       2283
Wright High School       1800
Name: Student ID, dtype: int64

In [14]:
# Total school budget

total_school_budget = school_data.set_index('school_name')['budget']

In [15]:
# per student budget

budget_per_student = (school_data.set_index('school_name')['budget']/school_data.set_index('school_name')['size'])

In [16]:
# Average Math Score

average_math_score = round(school_name['math_score'].mean(), 2)

In [17]:
# Average Reading Score

average_reading_score = round(school_name['reading_score'].mean(), 2)

In [18]:
# % Passing Math

pass_math_percent = round(school_data_complete[school_data_complete['math_score'] >= 70].groupby('school_name')['Student ID'].count()/total_student*100, 2)

In [19]:
# % Passing Reading

pass_read_percent = round(school_data_complete[school_data_complete['reading_score'] >= 70].groupby('school_name')['Student ID'].count()/total_student*100, 2)

In [20]:
# % Overall Passing (The percentage of students that passed math and reading.

overall_pass = round(school_data_complete[(school_data_complete['reading_score'] >= 70) & (school_data_complete['math_score'] >= 70)].groupby('school_name')['Student ID'].count()/total_student*100, 2)

In [21]:
school_summary = pd.DataFrame({"School Type"          : school_type,
                               "Total Students"       : total_student,
                               "Per Student Budget"   : budget_per_student,
                               "Total School Budget"  : total_school_budget,
                               "Average Math Score"   : average_math_score,
                               "Average Reading Score": average_reading_score,
                               '% Passing Math'       : pass_math_percent,
                               '% Passing Reading'    : pass_read_percent,
                               "% Overall Passing"    : overall_pass})

# formatting
school_summary.style.format({'Total Students': '{:}',
                             "Total School Budget": "${:,.2f}",
                             "Per Student Budget": "${:.2f}",
                             'Average Math Score': "{:6f}", 
                             'Average Reading Score': "{:6f}", 
                             "% Passing Math": "{:6f}", 
                             "% Passing Reading": "{:6f}"})

## Highest-Performing Schools (by % Overall Passing)

- Sort and display the top five performing schools by % overall passing.

In [22]:
# Sort the schools by `% Overall Passing` in descending order and display the top 5 rows.
top_schools = school_summary.sort_values("% Overall Passing", ascending = False)
top_schools.head(5).style.format({'Total Students': '{:}',
                                  "Total School Budget": "${:,.2f}", 
                                  "Per Student Budget": "${:.2f}", 
                                  "% Passing Math": "{:6f}", 
                                  "% Passing Reading": "{:6f}", 
                                  "% Overall Passing": "{:6f}"})

## Bottom Performing Schools (By % Overall Passing)

- Sort and display the five worst-performing schools by % overall passing.

In [23]:
# Sort the schools by `% Overall Passing` in ascending order and display the top 5 rows.
bottom_schools = top_schools.tail()
bottom_schools = top_schools.sort_values('% Overall Passing', ascending=True)

bottom_schools.head(5).style.format({'Total Students': '{: }',
                                     "Total School Budget": "${:,.2f}",
                                     "Per Student Budget": "${:.2f}",
                                     "% Passing Math": "{:6f}",
                                     "% Passing Reading": "{:6f}",
                                     "% Overall Passing": "{:6f}"})

## Math Scores by Grade

- Create a table that lists the average MATH Score for students of each grade level (9th, 10th, 11th, 12th) at each school.

In [24]:
# Use the code provided to separate the data by grade

ninth_graders    = school_data_complete[(school_data_complete["grade"] == "9th")]
tenth_graders    = school_data_complete[(school_data_complete["grade"] == "10th")]
eleventh_graders = school_data_complete[(school_data_complete["grade"] == "11th")]
twelfth_graders  = school_data_complete[(school_data_complete["grade"] == "12th")]


# Group by "school_name" and take the mean of each.

ninth_graders_scores    = ninth_graders.groupby('school_name')
tenth_graders_scores    = tenth_graders.groupby('school_name')
eleventh_graders_scores = eleventh_graders.groupby('school_name')
twelfth_graders_scores  = twelfth_graders.groupby('school_name')


# Use the code to select only the `math_score`.

ninth_grade_math_scores     = ninth_graders_scores["math_score"].mean()
tenth_grader_math_scores    = tenth_graders_scores["math_score"].mean()
eleventh_grader_math_scores = eleventh_graders_scores["math_score"].mean()
twelfth_grader_math_scores  = twelfth_graders_scores["math_score"].mean()

# Combine each of the scores above into single DataFrame called `math_scores_by_grade`
math_scores_by_grade = pd.DataFrame({"9th"  : ninth_grade_math_scores,
                                     "10th" : tenth_grader_math_scores,
                                     "11th" : eleventh_grader_math_scores,
                                     "12th" : twelfth_grader_math_scores })

# Display the DataFrame
math_scores_by_grade

9th       10th       11th       12th
school_name                                                      
Bailey High School     77.083676  76.996772  77.515588  76.492218
Cabrera High School    83.094697  83.154506  82.765560  83.277487
Figueroa High School   76.403037  76.539974  76.884344  77.151369
Ford High School       77.361345  77.672316  76.918058  76.179963
Griffin High School    82.044010  84.229064  83.842105  83.356164
Hernandez High School  77.438495  77.337408  77.136029  77.186567
Holden High School     83.787402  83.429825  85.000000  82.855422
Huang High School      77.027251  75.908735  76.446602  77.225641
Johnson High School    77.187857  76.691117  77.491653  76.863248
Pena High School       83.625455  83.372000  84.328125  84.121547
Rodriguez High School  76.859966  76.612500  76.395626  77.690748
Shelton High School    83.420755  82.917411  83.383495  83.778976
Thomas High School     83.590022  83.087886  83.498795  83.497041
Wilson High School     83.085578  83.724422  83.195326  83.035794
Wright High School     83.264706  84.010288  83.836782  83.644986

## Reading Score by Grade 

- Perform the same operations as above for reading scores.

In [25]:
# Use the code provided to separate the data by grade
ninth_graders    = school_data_complete[(school_data_complete["grade"] == "9th")]
tenth_graders    = school_data_complete[(school_data_complete["grade"] == "10th")]
eleventh_graders = school_data_complete[(school_data_complete["grade"] == "11th")]
twelfth_graders  = school_data_complete[(school_data_complete["grade"] == "12th")]


# Group by "school_name" and take the mean of each.
ninth_graders_scores    = ninth_graders.groupby('school_name')
tenth_graders_scores    = tenth_graders.groupby('school_name')
eleventh_graders_scores = eleventh_graders.groupby('school_name')
twelfth_graders_scores  = twelfth_graders.groupby('school_name')


# Use the code to select only the `read_score`.
ninth_grade_read_scores     = ninth_graders_scores["reading_score"].mean()
tenth_grader_read_scores    = tenth_graders_scores["reading_score"].mean()
eleventh_grader_read_scores = eleventh_graders_scores["reading_score"].mean()
twelfth_grader_read_scores  = twelfth_graders_scores["reading_score"].mean()

# Combine each of the scores above into single DataFrame called `read_scores_by_grade`
read_scores_by_grade = pd.DataFrame({"9th"  : ninth_grade_read_scores,
                                     "10th" : tenth_grader_read_scores,
                                     "11th" : eleventh_grader_read_scores,
                                     "12th" : twelfth_grader_read_scores })

# Display the DataFrame
read_scores_by_grade

9th       10th       11th       12th
school_name                                                      
Bailey High School     81.303155  80.907183  80.945643  80.912451
Cabrera High School    83.676136  84.253219  83.788382  84.287958
Figueroa High School   81.198598  81.408912  80.640339  81.384863
Ford High School       80.632653  81.262712  80.403642  80.662338
Griffin High School    83.369193  83.706897  84.288089  84.013699
Hernandez High School  80.866860  80.660147  81.396140  80.857143
Holden High School     83.677165  83.324561  83.815534  84.698795
Huang High School      81.290284  81.512386  81.417476  80.305983
Johnson High School    81.260714  80.773431  80.616027  81.227564
Pena High School       83.807273  83.612000  84.335938  84.591160
Rodriguez High School  80.993127  80.629808  80.864811  80.376426
Shelton High School    84.122642  83.441964  84.373786  82.781671
Thomas High School     83.728850  84.254157  83.585542  83.831361
Wilson High School     83.939778  84.021452  83.764608  84.317673
Wright High School     83.833333  83.812757  84.156322  84.073171

## Scores by School Spending

- Create a table that breaks down school performances based on average Spending Ranges (Per Student). Use 4 reasonable bins to group school spending. Include in the table each of the following:
    - Average Math Score
    - Average Reading Score
    - % Passing Math
    - % Passing Reading
    - Overall Passing Rate (Average of the above two)

In [26]:
# Establish the bins 
spending_bins = [0, 585, 630, 645, 680]
labels        = ["<$585", "$585-630", "$630-645", "$645-680"]

In [27]:
# Create a copy of the school summary since it has the "Per Student Budget"

school_data_complete['spending_spending_bins'] = pd.cut(school_data_complete['budget']/school_data_complete['size'], spending_bins, labels = labels)

#group by spending
school_spending_df = school_data_complete.groupby('spending_spending_bins')

In [28]:
# Use `pd.cut` to categorize spending based on the bins.
school_data_complete["Spending Ranges (Per Student)"] = pd.cut(school_data_complete['budget'] / school_data_complete['size'],
                                                               bins = spending_bins, labels = labels)

school_spending_df = school_data_complete.groupby('Spending Ranges (Per Student)')

In [29]:
#  Calculate averages for the desired columns. 
spending_math_scores     = school_spending_df['math_score'].mean()
spending_reading_scores  = school_spending_df['reading_score'].mean()
spending_passing_math    = school_data_complete[school_data_complete['math_score'] >= 70].groupby('spending_spending_bins')['Student ID'].count()/school_spending_df['Student ID'].count()*100
spending_passing_reading = school_data_complete[school_data_complete['reading_score'] >= 70].groupby('spending_spending_bins')['Student ID'].count()/school_spending_df['Student ID'].count()*100
overall_passing_spending = school_data_complete[(school_data_complete['reading_score'] >= 70) & (school_data_complete['math_score'] >= 70)].groupby('spending_spending_bins')['Student ID'].count()/school_spending_df['Student ID'].count()*100

spending_summary = pd.DataFrame({"Average Math Score": spending_math_scores,
                                 "Average Reading Score": spending_reading_scores,
                                 "% Passing Math": spending_passing_math,
                                 "% Passing Reading": spending_passing_reading,
                                 "% Overall Passing": overall_passing_spending})
            
# reorder columns
spending_summary = spending_summary[["Average Math Score",
                                     "Average Reading Score",
                                     "% Passing Math",
                                     "% Passing Reading",
                                     "% Overall Passing"]]

# formating
spending_summary.style.format({'Average Math Score': '{:.2f}', 
                              'Average Reading Score': '{:.2f}', 
                              '% Passing Math': '{:.2f}', 
                              '% Passing Reading':'{:.2f}', 
                              '% Overall Passing': '{:.2f}'})

## Scores by School Size

- Perform the same operations as above, based on school size.

In [30]:
# Establish the bins.
size_bins = [0, 1000, 2000, 5000]
labels    = ["Small (<1000)", "Medium (1000-2000)", "Large (2000-5000)"]

In [31]:
# Categorize the spending based on the bins
# Use `pd.cut` on the "Total Students" column of the `per_school_summary` DataFrame.

school_data_complete["School Size"] = pd.cut(school_data_complete['size'], bins=size_bins, labels = labels)
school_size_df                      = school_data_complete.groupby('School Size')

In [32]:
# Calculate averages for the desired columns. 
size_math_scores     = school_size_df['math_score'].mean()
size_reading_scores  = school_size_df['math_score'].mean()
size_passing_math    = school_data_complete[school_data_complete['math_score'] >= 70].groupby('School Size')['Student ID'].count()/school_size_df['Student ID'].count()*100
size_passing_reading = school_data_complete[school_data_complete['reading_score'] >= 70].groupby('School Size')['Student ID'].count()/school_size_df['Student ID'].count()*100
size_overall_passing = school_data_complete[(school_data_complete['reading_score'] >= 70) & (school_data_complete['math_score'] >= 70)].groupby('School Size')['Student ID'].count()/school_size_df['Student ID'].count()*100

In [33]:
# Create a DataFrame called `size_summary` that breaks down school performance based on school size (small, medium, or large).
# Use the scores above to create a new DataFrame called `size_summary`
size_summary = pd.DataFrame({"Average Math Score"     : size_math_scores,
                             "Average Reading Score"  : size_reading_scores,
                             '% Passing Math'         : size_passing_math,
                             '% Passing Reading'      : size_passing_reading,
                             '% Overall Passing'      : size_overall_passing})

size_summary.style.format({'Average Math Score'       : '{:.6f}', 
                              'Average Reading Score' : '{:.6f}', 
                              '% Passing Math'        : '{:.6f}', 
                              '% Passing Reading'     :'{:.6f}', 
                              '% Overall Passing'     : '{:.6f}'})

## Scores by School Type

- Perform the same operations as above, based on school type.

In [34]:
schoo_type = school_data_complete.groupby("type")

# calculations 
average_math_score_by_type              = schoo_type['math_score'].mean()
average_reading_score_by_type           = schoo_type['math_score'].mean()
average_percent_passing_math_by_type    = school_data_complete[school_data_complete['math_score'] >= 70].groupby('type')['Student ID'].count()/schoo_type['Student ID'].count()*100
average_percent_passing_reading_by_type = school_data_complete[school_data_complete['reading_score'] >= 70].groupby('type')['Student ID'].count()/schoo_type['Student ID'].count()*100
average_percent_overall_passing_by_type = school_data_complete[(school_data_complete['reading_score'] >= 70) & (school_data_complete['math_score'] >= 70)].groupby('type')['Student ID'].count()/schoo_type['Student ID'].count()*100

In [35]:
# Assemble the new data by type into a DataFrame called `type_summary`

type_summary = pd.DataFrame({"Average Math Score"    : average_math_score_by_type,
                             "Average Reading Score" : average_reading_score_by_type,
                             '% Passing Math'        : average_percent_passing_math_by_type,
                             '% Passing Reading'     : average_percent_passing_reading_by_type,
                             "% Overall Passing"     : average_percent_overall_passing_by_type})


# Display results
type_summary.style.format({'Average Math Score'       : '{:.6f}', 
                              'Average Reading Score' : '{:.6f}', 
                              '% Passing Math'        : '{:.6f}', 
                              '% Passing Reading'     :'{:.6f}', 
                              '% Overall Passing'     : '{:.6f}'})

## Observable Trends
1. The dataset encompasses a total of `15` schools, `39,170` students. The total students average reading score is `81.87` which is higher than the average math score 78.98. When we see the passing rate reading `(85.80%)` has a higher rate than math `(74.98 %)`, the overall passing rate is `65.17%`. These findings indicate that student's score more in reading than math.


2. The school summary finding showed that from `15` schools `7` of them are found at the district level, and the other `8` of them found at the charter level. The number of students in schools ranged between `427` students (Holden High School) to `4976` students (Bailey High School). The general finding also showed that district schools have more students than charter schools. It is legitimate to conclude that district schools have more students than charter schools because of their geographical coverage.


3. The total budget field of the school summery showed the amount of money each school consumed. The maximum budget is `$3,124,928.00` its Bailey High School, and the lower budget is `$248,087.00` its Holden High School. We can see later Bailey High School is not in the top performing schools. The schools budget allocaters,and Bailey High School should to be work on this to improve the quality. Regarding budget per student, Huang High School `($655.000)` has a higher budget per student, and Wilson High School `($578.00)` allocates a lower budget per student. Huang High School is one of the bottom performing school the same as Bailey High School it one area for future improvement to see education quality as it budgetd. The total district schools budget `($17,347,923.00)` is higher than Charter schools budget `($7,301,505.00)`, however, when we come to the total allocation of Budget Per Student charter schools allocation `($4,796.00)` is greater than district schools `($4,505.00)`. This may due to charter schools have a smaller number of students.


4. The findings from the data showed that in all metrics (average, percent, and overall) of reading and math scores Charter schools have a greater score than District schools. Pena High School has the maximum `(83.839917)` average math score `(83.839917)`, percent of students passing math `(94.59%)` and average reading score `(84.044699)`. This implies that Pena High School has a good quality, and reputation in teaching math, and reading. On the other side, Huang High School has a lower average math score `(76.842711)`, and percent of students passing math `(65.683922)`. Regarding the percent of passing reading class Thomas High School has a higher percentage value `(97.30%)`. On the flip side, Rodriguez High School has a lower Average Reading Score `(80.744686)`, and Ford High School has a lower passing reading. In the overall percentage passing Cabrera High School has the maximum passing percentage rate `(91.33 %)`, and Rodriguez High School has a lower percentage of `(52.98)` passing.


5. Grade 11 students at Holden High School holds a higher average Math score `(85)` comparing all schools and grades. Holden High School grade `12` students possess a higher average reading score `(84.698795)`.


6. The other irregular finding of this analysis is that when the budget per student increases student's average math score, the average reading score, percent of passing math, percent of passing reading, and the overall passing decreases. This finding commences investigating the real motivators for academic achievement, and give an open space for researchers to work on this more.


7. The data revealed that when the number of students sizes higher in schools the average math scores, average reading scores, passing math, passing reading, and overall passing decrease. So, we can infer that, when student size rises, it negatively influences student achievement. Therefore, schools, cities, districts, and charters should work on student size in schools, and classes because it directly affects student achievement for so many reasons.